### PR TITLE
Adding Surface Magnetism Solver + Hand Ray Example scene

### PR DESCRIPTION
**Pull request by cre8ivepark**
**NOTE:** Dependent on #6304 PR

MRTK_SurfaceMagExample
## Overview
Currently, there is no easy way to make an object attached to the endpoint of the Hand Ray and make it placed on a surface with the surface normal aligned. Adding a new example scene for the Spatial Awareness + Surface Magnetism

## Changes
In SurfaceMagnetism.cs, I added Hand Ray option. However, **I am not sure if this is the right place to add since the section name is 'Raycasting Direction Mode'. I feel this could go into SolverHandler's Tracked Target Object. Need help here.** 

                    case RaycastDirectionMode.HandRay:
                        PointerUtils.TryGetPointerEndpoint<IMixedRealityPointer>(Handedness.Both, InputSourceType.Hand, out endPoint);
                        break;

2019-10-13 12_51_48-Window
2019-10-13 13_04_49-Window

## Example Scene
Added a new example scene **SurfaceMagnetismSpatialAwarenessExample** to demonstrate a slate following the Spatial Mapping surface. 
2019-10-13 17_36_30-Unity 2018 4 7f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK

- Fixes: #6279

### CODE REVIEW

**Comment by cre8ivepark** [comment]
@Troy-Ferrell the Hand joint options make the object attach to the hand joints, not to the endpoint of the ray. I couldn't find a wat to achieve it. However, for 'Head' option, it already gives me the capability to attach object to the endpoint of the gaze, combined with Surface Magnetism's Raycast Direction

**Comment by Troy-Ferrell** [comment]
> Hand

When this property is set as such on Surface Magnetism, it will use the forward ray direction of the tracked target type on the associated SolverHandler.
Current Raycast Direction Mode = Tracked Target Forward

In the case of head, it is the camera's forward gaze.
In the case of the hand, it will be the forward vector of the palm (or whatever joint selected) transform. 

So it actually should work. I swear I was using it right before GA. 

I'll look into this tomorrow but the current API should allow what you are trying to do @cre8ivepark

**Comment by cre8ivepark** [comment]
@Troy-Ferrell That would be amazing! I'll also check tonight with palm + tracked target forward option.

**Comment by Troy-Ferrell** [comment]
> @Troy-Ferrell That would be amazing! I'll also check tonight with palm + tracked target forward option.

Actually I already checked. Sorry, that is what I meant by things were currently broken and why I updated your bug title. For some reason, the forward vector on the hand palm join (NOTE at least in input simulation) was not working with the Surface Magnetism (TrackedTargetForward option).

I'm gonna try to debug later today

**Comment by cre8ivepark** [comment]
One more question: will Palm + forward raycast give us the same result as Hand Ray's endpoint? Because the Hand Ray's ray is coming from the shoulder, the target object position might not perfectly match the endpoint of the hand ray?

**Comment by Troy-Ferrell** [comment]
> One more question: will Palm + forward raycast give us the same result as Hand Ray's endpoint? Because the Hand Ray's ray is coming from the shoulder, the target object position might not perfectly match the endpoint of the hand ray?

I figured out why hand joint was failing. But you bring up a good point. It wouldn't exactly be the same hand ray as used by the pointers. 

I think I have a solution though that should fit nicely into the existing framework to fix the bug and be a bit cleaner code than what is in SolverHandler. I'll create a PR for that soon.

@cre8ivepark , we should still keep this PR though. You did a fantastic job helping with the Interactable scene update. It would be incredible if you could apply your same magic to the Solver example scene.

I'll sync with you soon to talk about it.

SM_Fix

**Comment by cre8ivepark** [comment]
@Troy-Ferrell Looks amazing! Once you open the PR, I'll add the Spatial Awareness + Surface Magnetism example scene.

**Comment by Troy-Ferrell** [comment]
@cre8ivepark, pull or wait for this branch and then update your PR
#6304

**Comment by cre8ivepark** [comment]
@Troy-Ferrell 
Reverted SurfaceMagnetism.cs
Updated the scene example to use 'Controller Ray' as a tracked target type.
Created new profiles for the example scene. (to enable Spatial Awareness)
Tested on HoloLens 2 device.
2019-10-28 13_12_27-MixedRealityToolkit-Unity - Microsoft Visual Studio

File: Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset [hunk 1]
```diff
@@ -0,0 +1,54 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
+--- !u!114 &11400000
+MonoBehaviour:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset]: Is there anything special here other than turning on SA? If not, I'd say let's just re-use "DefaultMixedRealityToolkitConfigurationProfile".

Comment (cre8ivepark) [reply Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset]: In this profile, I made the mesh invisible to reduce visual noise.

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset]: I'd say use the default and then add hand menu or toggle menu to switch out between occluded vs visible states in SA system

Comment (cre8ivepark) [reply Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleMRTKConfigurationProfile.asset]: Great idea!

File: Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleSpatialAwarenessSystemProfile.asset [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+--- !u!114 &11400000
+MonoBehaviour:
+  m_ObjectHideFlags: 0
+  m_CorrespondingSourceObject: {fileID: 0}
+  m_PrefabInstance: {fileID: 0}
+  m_PrefabAsset: {fileID: 0}
+  m_GameObject: {fileID: 0}
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 19f279aded72cb741b4de89a54359dd4, type: 3}
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleSpatialAwarenessSystemProfile.asset]: Will you be removing this asset, per Troy's comment?

Comment (cre8ivepark) [reply Assets/MixedRealityToolkit.Examples/Demos/Solvers/Profiles/SurfaceMagnetismExampleSpatialAwarenessSystemProfile.asset]: Yes. And adding toggle button for spatial mesh visual.

**Comment by cre8ivepark** [comment]
Doing some final device testings for
- Spatial Mesh visual toggle button
- Added another plane example object with tracked object 'head' option for HoloLens 1st gen users.

**Comment by cre8ivepark** [comment]
Spatial Mesh visual toggle button
2019-10-31 16_15_37-Unity 2018 4 6f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK

Added another plane example object with tracked object 'head' option for HoloLens 1st gen users.
2019-10-31 16_15_26-Unity 2018 4 6f1 Personal - SurfaceMagnetismSpatialAwarenessExample unity - MRTK

Deleted custom profile for the spatial awareness

**Pull request merged**